stream/stream_dvbv5.c

```diff
@@ -0,0 +1,687 @@
+        if (parms->current_sys == SYS_ATSC &&
+            entry->props[i].cmd == DTV_MODULATION) {
+            if (data != VSB_8 && data != VSB_16)
+                dvb_fe_store_parm(parms,
+                          DTV_DELIVERY_SYSTEM,
+                          SYS_DVBC_ANNEX_B);
+        }
+    }
+
+    dvb_file_free(dvb_file);
+    return 0;
+}
+
+/*
+ * This function was adapted for MPV from v4l-utils/utils/dvb/dvbv5-zap.c
+ * version v4l-utils-1.20.0-118-g9a628fbc
+ */
+static int setup_frontend(stream_t *stream, struct dvb_v5_fe_parms *parms)
+{
+    int rc;
+
+    rc = dvb_fe_set_parms(parms);
+    if (rc < 0) {
+        MP_ERR(stream, "dvb_fe_set_parms failed\n");
+        return -1;
+    }
+
+    return 0;
+}
+
+/*
+ * This function was adapted for MPV from v4l-utils/utils/dvb/dvbv5-zap.c
+ * version v4l-utils-1.20.0-118-g9a628fbc
+ */
+static int check_frontend(stream_t *stream, struct dvb_v5_fe_parms *parms)
+{
+    int rc;
+    fe_status_t status = 0;
+    do {
+        rc = dvb_fe_get_stats(parms);
+        if (rc) {
+            MP_ERR(stream, "dvb_fe_get_stats failed\n");
+            usleep(1000000);
+            continue;
+        }
+
+        status = 0;
+        rc = dvb_fe_retrieve_stats(parms, DTV_STATUS, &status);
+        if (status & FE_HAS_LOCK)
+            break;
+        usleep(1000000);
+    } while (!timeout_flag);
+
+    return status & FE_HAS_LOCK;
+}
+
+/*
+ * This function was adapted for MPV from v4l-utils/utils/dvb/dvbv5-zap.c
+ * version v4l-utils-1.20.0-118-g9a628fbc
+ */
+static int dvbv5_open(stream_t *stream)
+{
+    pthread_t timeout;
+    char *demux_dev, *dvr_dev;
+    int lnb = -1;
+    int vpid = -1, apid = -1, sid = -1;
+    int pmtpid = 0;
+    struct dvb_open_descriptor *pat_fd = NULL, *pmt_fd = NULL;
+    struct dvb_open_descriptor *sdt_fd = NULL;
+    struct dvb_open_descriptor *sid_fd = NULL, *dvr_fd = NULL;
+    struct dvb_open_descriptor *audio_fd = NULL, *video_fd = NULL;
+    int r, ret;
+    struct dvb_v5_fe_parms *parms = NULL;
+    struct dvb_device *dvb;
+    struct dvb_dev_list *dvb_dev;
+    static dvb_priv_t cached_priv = {0};
+    dvb_priv_t *priv = NULL;
+    struct m_config_cache *opts_cache;
+    dvbv5_opts_t *opts;
+
+    opts_cache = m_config_cache_alloc(stream, stream->global, &stream_dvbv5_conf);
+    opts = opts_cache->opts;
+
+    if (opts->cfg_keep_tune && cached_priv.dvr_fd) {
+        stream->fill_buffer = dvbv5_streaming_read;
+        stream->close = dvbv5_close;
+        stream->control = dvbv5_stream_control;
+        stream->streaming = true;
+        stream->demuxer = "lavf";
+        stream->lavf_type = "mpegts";
+        stream->priv = talloc_memdup(stream, &cached_priv, sizeof(dvb_priv_t));
+        priv = stream->priv;
+        priv->opts_cache = opts_cache;
+        priv->opts = opts;
+        dvb = priv->dvb;
+        dvr_dev = priv->dvr_dev;
+        demux_dev = priv->demux_dev;
+        audio_fd = priv->audio_fd;
+        video_fd = priv->video_fd;
```

Comment:
Most of these lines are duplicated later, after the `if` is closed, and might be better kept before the `if` to prevent duplication.